### ISSUE TITLE
Change resizing algorithm to lanczos or at least cubic

### ISSUE THREAD

**Issue by Beep6581** [LuminanceHDR/LuminanceHDR#8]
The resizing algorithm used for the preview results in horrible jaggies. Please change it to Lanczos or at least (bi)cubic.

**Comment by Beep6581** [comment]
The preview quality is still low in 2.5.0.

![imgur-2017_04_19-14 41 10](https://cloud.githubusercontent.com/assets/5844619/25180440/47af772a-250e-11e7-8c3d-aac2e7383010.png)
![imgur-2017_04_19-14 41 00](https://cloud.githubusercontent.com/assets/5844619/25180441/47b0010e-250e-11e7-9179-a5f8cdcd33d1.png)

**Comment by Beep6581** [comment]
@fcomida looks like the Qt way of doing this is by enabling the "Aliasing" render hint.
http://doc.qt.io/qt-5/qpainter.html#RenderHint-enum
This could also help:
http://stackoverflow.com/questions/17734290/antialiasing-not-working-in-qgraphicsview

**Comment by fcomida** [comment]
Thank you for this tip, pushed right now, try it out.

**Comment by Beep6581** [comment]
@fcomida great! While it's compiling, I wanted to point out that:
> QPainter::HighQualityAntialiasing	0x08	This value is obsolete and will be ignored, use the Antialiasing render hint instead.

**Comment by fcomida** [comment]
@Beep6581 It didn't complain on Qt 5.6, thank you.

**Comment by Beep6581** [comment]
@fcomida the docs say the value will be ignored.

* Downscaling test
  * Without patch:
![imgur-2017_04_29-20 02 16](https://cloud.githubusercontent.com/assets/5844619/25557695/cec11aba-2d16-11e7-8dd3-7cb32c800e10.png)
  * With patch:
![imgur-2017_04_29-20 02 11](https://cloud.githubusercontent.com/assets/5844619/25557696/d4330832-2d16-11e7-93f2-997c32ad10aa.png)
* Upscaling test
  * Without patch:
![imgur-2017_04_29-20 00 14](https://cloud.githubusercontent.com/assets/5844619/25557683/7e424eec-2d16-11e7-9312-b2b115b4b374.png)
  * With patch:
![imgur-2017_04_29-20 00 19](https://cloud.githubusercontent.com/assets/5844619/25557684/84851bc2-2d16-11e7-9782-a5f26e642175.png)

The results are consistent with expectations: the preview is smoother.

Bug now it's clear that there is a second area which needs improvements - the resizing algorithm. Not the one which resizes the preview, but the one which resizes the actual image.

I set the result size to 512x343:
![imgur-2017_04_29-20 10 00](https://cloud.githubusercontent.com/assets/5844619/25557745/e37f173a-2d17-11e7-98a8-51f52ee932fe.png)
It looks like no interpolation was used.
Here is a comparison of the LHDR resized image vs ImageMagick's Lanczos:
![compare](https://cloud.githubusercontent.com/assets/5844619/25557763/6515d9fa-2d18-11e7-8979-3baeb0fbd9e9.png)

@fcomida could you take a look at that too? It's very important as it severely affects every image not saved at its original full size.

**Comment by fcomida** [comment]
@Beep6581 I will, indeed those Qt settings are only for displaying the images, for saving and for the cropping/resizing features an interpolation algorithm must be implemented. I was asked to implement dithering too, was he you?

**Comment by Beep6581** [comment]
@fcomida no, I did not ask for dithering. Maybe when whoever asked for dithering sees good interpolation they won't want it anymore.

RawTherapee code is very fast and the resizing is great, if you'd like to copypasta: https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/ipresize.cc

**Comment by fcomida** [comment]
![screenshot_20170430_025930](https://cloud.githubusercontent.com/assets/11018726/25560331/66220bd2-2d51-11e7-81d9-218c2d259239.png)
![screenshot_20170430_030020](https://cloud.githubusercontent.com/assets/11018726/25560332/6622e7d2-2d51-11e7-941e-540443647c12.png)
@Beep6581 Bilinear, Lanczos

**Comment by Beep6581** [comment]
@fcomida bottom one is infinitely better!

**Comment by fcomida** [comment]
![screenshot_20170430_131219](https://cloud.githubusercontent.com/assets/11018726/25563907/ee15803c-2da6-11e7-90a1-0298e1894dc6.png)
![screenshot_20170430_131333](https://cloud.githubusercontent.com/assets/11018726/25563908/ee1765b4-2da6-11e7-8b57-0333ccbd1505.png)
@Beep6581 It can produce artifacts though, at very low resolutions.

**Comment by Beep6581** [comment]
The image resizing quality before this patch was unusably bad. The overall improvement is terrific, but those artifacts are a problem. I reproduced them using 078314eb816c29450cc7185ceb5051799554a16d at any resolution, along very high contrast edges. Some kind of overshoot in the calculation when a low value lies directly adjacent to a very high value?

What color space is the LHDR image in when its resized?

**Comment by Beep6581** [comment]
![imgur-2017_04_30-17 33 59](https://cloud.githubusercontent.com/assets/5844619/25565657/6f12948a-2dcb-11e7-869d-db0dcabbc252.png)

**Comment by Beep6581** [comment]
I had a chat with a person who knows things, and the summary is that:
* Downscaling performed on HDR images needs non-negative kernels
* Lanczos shouldn't be used on HDR files, it is inherently unsuitable for them. LHDR should use bilinear with box sampling for that:
> bilinear is only useful for things between 0.5x and 1x for downsampling, so you need to do box sampling first, to do integer scaling, and then you do bilinear

Code here: https://github.com/CarVac/filmulator-gui/blob/master/filmulator-gui/core/scale.cpp

**Comment by fcomida** [comment]
@Beep6581 our previous code is derived from here: [bilinear image scaling](http://tech-algorithm.com/articles/bilinear-image-scaling/) with added OpenMP support and block based resampling.

**Comment by CarVac** [comment]
@fcomida Did you implement the "Caveat" in that article?

>Bilinear scaling performs best when the desired output dimension is no more than double or half its original size. If that is the case however, it might be good to implement additional technique called Mip Mapping on top of the existing algorithm.

You may be properly implementing bilinear scaling, but first you have to get it within a factor of two of the desired size using integer scaling.

**Comment by CarVac** [comment]
If this is the function in question:

https://github.com/LuminanceHDR/LuminanceHDR/blob/master/src/Libpfs/manip/resize.hxx#L214

It looks like it doesn't perform any of the necessary integer scaling.

**Comment by fcomida** [comment]
The function is that one, it was implemented by Davide time ago. So no Lanczos at all? In my limited tests it performs well except for large scaling factors.

**Comment by CarVac** [comment]
Lanczos is fine if you do resizing after tonemapping, which is not what you care about since you want to do the tonemapping on a lower resolution so it can run faster.

Lanczos has problems with ringing when the initial contrast is too high. For HDR content like this, since the highlights are tremendously brighter than their surroundings, that can result in the very visible halos that @Beep6581 showed in his latest screenshot.

The bilinear algorithm looks correct, but it's currently bilinearly interpolating between only the four neighbors nearest next to each destination pixel, not far off from plain nearest-neighbor. It needs to be integer downsampled to a size between 1X and 2X the end result size before you run the bilinear filter.

**Comment by fcomida** [comment]
@CarVac Yes, I was actually me that pointed out those artifacts. I'm gonna improve bilinear then. At small scaling factors Lanczos doesn't seem to produce artifacts though, have I been lucky with my testing? Can we switch among the two (Lanczos for small scaling factor an then bilinear)?

**Comment by CarVac** [comment]
In my opinion, I think you should stick to bilinear only, to ensure that artifacts don't occur.

**Comment by Beep6581** [comment]
@fcomida in the library example image set, I got artifacts along high contrast edges even for small scaling factors.

**Comment by fcomida** [comment]
@Beep6581 @CarVac got it. Goodbye Lancozs....
